From c62bd6bca2d4902f1bd4a57c22e4bd7c674c3d0e Mon Sep 17 00:00:00 2001
From: Alon Levy <alevy@redhat.com>
Date: Fri, 13 Jan 2012 12:54:50 +0200
Subject: [PATCH 32/46] client/windows: fix typo, make error messages unique

---
 client/windows/platform_utils.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/client/windows/platform_utils.cpp b/client/windows/platform_utils.cpp
index eb87468..83fd383 100644
--- a/client/windows/platform_utils.cpp
+++ b/client/windows/platform_utils.cpp
@@ -48,7 +48,7 @@ HBITMAP get_bitmap_res(int id)
 {
     HBITMAP bitmap = LoadBitmap(GetModuleHandle(NULL), MAKEINTRESOURCE(id));
     if (!bitmap) {
-        THROW("get bitmpa #%d failed", id);
+        THROW("get bitmap #%d failed", id);
     }
     return bitmap;
 }
@@ -58,7 +58,7 @@ HBITMAP get_alpha_bitmap_res(int id)
     AutoGDIObject bitmap(LoadImage(GetModuleHandle(NULL), MAKEINTRESOURCE(id), IMAGE_BITMAP, 0, 0,
                                    LR_DEFAULTCOLOR | LR_CREATEDIBSECTION | LR_SHARED));
     if (!bitmap.valid()) {
-        THROW("get bitmpa #%d failed", id);
+        THROW("get alpha bitmap #%d failed", id);
     }
 
     BITMAP src_info;
-- 
1.7.7.6

